src/java/org/apache/cassandra/index/sai/utils/RangeIntersectionIterator.java

```diff
@@ -0,0 +1,229 @@
+                        alreadyAvanced = index;
+                        // This iterator jumped over, so the other iterators are lagging behind now,
+                        // including the ones already advanced in the earlier cycles of the inner loop.
+                        // Therefore, restart the inner loop in order to advance
+                        // the other iterators except this one to match the new highest key.
+                        continue outer;
+                    }
+                    assert nextKey.compareTo(highestKey) == 0:
+                    String.format("skipTo skipped to an item smaller than the target; " +
+                                  "iterator: %s, target key: %s, returned key: %s", range, highestKey, nextKey);
+                }
+            }
+            // If we reached here, next() has been called at least once on each range iterator and
+            // the last call to next() on each iterator returned a value equal to the highestKey.
+            return highestKey;
+        }
+        return endOfData();
+    }
+
+    protected void performSkipTo(PrimaryKey nextToken)
+    {
+        for (RangeIterator range : ranges)
+            if (range.hasNext())
+                range.skipTo(nextToken);
+    }
+
+    /**
+     * Fetches the next available item from the iterator, such that the item is not lower than the given key.
+     * If no such items are available, returns null.
+     */
+    private PrimaryKey nextOrNull(RangeIterator iterator, PrimaryKey minKey)
+    {
+        iterator.skipTo(minKey);
+        return iterator.hasNext() ? iterator.next() : null;
+    }
+
+    public void close() throws IOException
+    {
+        ranges.forEach(FileUtils::closeQuietly);
+    }
+
+    public static Builder builder()
+    {
+        return new Builder();
+    }
+
+    public static Builder selectiveBuilder()
+    {
+        return selectiveBuilder(INTERSECTION_CLAUSE_LIMIT);
+    }
+
+    public static Builder selectiveBuilder(int limit)
+    {
+        return new Builder(limit);
+    }
+
+    public static class Builder extends RangeIterator.Builder
+    {
+        private final int limit;
+        protected List<RangeIterator> rangeIterators = new ArrayList<>();
+
+        public Builder()
+        {
+            this(Integer.MAX_VALUE);
+        }
+
+        public Builder(int limit)
+        {
+            super(IteratorType.INTERSECTION);
+            this.limit = limit;
+        }
+
+        public RangeIterator.Builder add(RangeIterator range)
+        {
+            if (range == null)
+                return this;
+
+            if (range.getCount() > 0)
+                rangeIterators.add(range);
+            else
+                FileUtils.closeQuietly(range);
+            statistics.update(range);
+
+            return this;
+        }
+
+        public RangeIterator.Builder add(List<RangeIterator> ranges)
```

Comment:
Do we need this override? Seems like just leaving the base class version in place works. It seems like this is also the case in `RangeUnionIterator`'s builder.